In [359]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [360]:
data=pd.read_csv('OnlineRetail.csv',encoding = "ISO-8859-1")
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [361]:
# just to know how many Unique values are there for object fields 

print ("Unique InvoiceNo",len(data.InvoiceNo.unique()))
print ("Unique StockCode",len(data.StockCode.unique()))


Unique InvoiceNo 25900
Unique StockCode 4070


In [362]:
data.shape

(541909, 8)

In [363]:
data.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [364]:
# To check and count Missing/Null data in all features 

print ("Is null data present:- ",data.isnull().values.any())

data.isnull().sum()

Is null data present:-  True


InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [365]:
data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [366]:
# Can be replaced missing CustomerID with Median value

print ("mean value of customID:", data.CustomerID.mean())          #  Mean value 
print ("Median value of customID:", data.CustomerID.median())        #  Median Value

#df = data
#df['CustomerID'] = df['CustomerID'].fillna(df['CustomerID'].median())


# Dropping CustomerID as of now to take decison on Customer category..
data = data[data['CustomerID'].notna()]

print ("Data dimension after dropping null customerID:",data.shape)

print ("Unique customid values:", len(data.CustomerID.unique()))   #  Checking Unique customId fields 


mean value of customID: 15287.690570239585
Median value of customID: 15152.0
Data dimension after dropping null customerID: (406829, 8)
Unique customid values: 4372


### From Above Unique values we can say that there are approximate <font color=red>4372</font> Unique Customers 
### dealing with the Online Retail Shop, 

* Around 135080 Customers doesn't have CustomerID which is like necessory information , hence customer ID is null we cant comment on their behaviour we are dropping them as of now from the analysis .

* we could replace them using median but it will Bias the decison.

* we can drop description column as its not much important column as its not Biasing the decison..


In [367]:
data.isnull().sum()
data.shape

(406829, 8)

In [368]:
data.drop('Description',axis=1,inplace=True)
data.shape

(406829, 7)

###  Analysis on InvoiceNo..

In [369]:
cancel_order_index = (data['InvoiceNo'].str.contains('^C.*'))
data['InvoiceNo'].str.contains('^C.*').value_counts()

False    397924
True       8905
Name: InvoiceNo, dtype: int64

In [370]:
cancelled_Orders  = data[cancel_order]
cancelled_Orders.head()
data.loc[data['InvoiceNo'] == '536391']

C:\Users\Jigar\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: [InvoiceNo, StockCode, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

### From above two cells we can observe that there are around <font color=red>8905</font> cancelled order, 
### also observed that cancelled order Invoice is not matched with successful order Invoiceno by removing Initial 'C' character.

In [371]:
data.head()

InvoiceNo StockCode  Quantity     InvoiceDate  UnitPrice  CustomerID  \
0    536365    85123A         6  12/1/2010 8:26       2.55     17850.0   
1    536365     71053         6  12/1/2010 8:26       3.39     17850.0   
2    536365    84406B         8  12/1/2010 8:26       2.75     17850.0   
3    536365    84029G         6  12/1/2010 8:26       3.39     17850.0   
4    536365    84029E         6  12/1/2010 8:26       3.39     17850.0   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom

In [372]:
data['Bill'] = data.Quantity * data.UnitPrice
data.head()


InvoiceNo StockCode  Quantity     InvoiceDate  UnitPrice  CustomerID  \
0    536365    85123A         6  12/1/2010 8:26       2.55     17850.0   
1    536365     71053         6  12/1/2010 8:26       3.39     17850.0   
2    536365    84406B         8  12/1/2010 8:26       2.75     17850.0   
3    536365    84029G         6  12/1/2010 8:26       3.39     17850.0   
4    536365    84029E         6  12/1/2010 8:26       3.39     17850.0   

          Country   Bill  
0  United Kingdom  15.30  
1  United Kingdom  20.34  
2  United Kingdom  22.00  
3  United Kingdom  20.34  
4  United Kingdom  20.34

In [373]:
# We can remove the unwanted Columns like, Country, UnitPrice, Quantity, StockCode from dataframe 

data.drop(['StockCode','Quantity','UnitPrice','Country'], axis=1, inplace=True)
data.head()

InvoiceNo     InvoiceDate  CustomerID   Bill
0    536365  12/1/2010 8:26     17850.0  15.30
1    536365  12/1/2010 8:26     17850.0  20.34
2    536365  12/1/2010 8:26     17850.0  22.00
3    536365  12/1/2010 8:26     17850.0  20.34
4    536365  12/1/2010 8:26     17850.0  20.34

### Can measure Aggregate Bills - like revenue generated from each customer Monthly 

In [374]:

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'],format='%m/%d/%Y %H:%M')
data['InvoiceDate'].min()

'''

data['InvoiceMonth'] =  data['InvoiceDate'].values.astype('datetime64[M]')
data.head()
print (data.shape)

# Dropping other non -related info or lets put things out of box like stock, country etc ..

df = data[['CustomerID', 'InvoiceMonth', 'Bill']].copy()
df.head()

monthly_bills = df.groupby(['CustomerID','InvoiceMonth'])[['Bill']].agg('sum').reset_index()
monthly_bills.head()

monthly_bills.shape
'''


"\n\ndata['InvoiceMonth'] =  data['InvoiceDate'].values.astype('datetime64[M]')\ndata.head()\nprint (data.shape)\n\n# Dropping other non -related info or lets put things out of box like stock, country etc ..\n\ndf = data[['CustomerID', 'InvoiceMonth', 'Bill']].copy()\ndf.head()\n\nmonthly_bills = df.groupby(['CustomerID','InvoiceMonth'])[['Bill']].agg('sum').reset_index()\nmonthly_bills.head()\n\nmonthly_bills.shape\n"

In [375]:
'''
# Customers are not continuous in buying the things .. we can see irregularities in by checking customers purchase ..
#e.g. 
monthly_bills[monthly_bills['CustomerID'] == 12347.0]
'''

"\n# Customers are not continuous in buying the things .. we can see irregularities in by checking customers purchase ..\n#e.g. \nmonthly_bills[monthly_bills['CustomerID'] == 12347.0]\n"

In [376]:
'''
final_df = monthly_bills[['CustomerID','Bill']].copy()
final_df.head()
'''

"\nfinal_df = monthly_bills[['CustomerID','Bill']].copy()\nfinal_df.head()\n"

In [377]:
'''
final_df.head()
final_df.groupby(['CustomerID'])[['Bill']].agg('sum').reset_index()
#final_df.head()
'''

"\nfinal_df.head()\nfinal_df.groupby(['CustomerID'])[['Bill']].agg('sum').reset_index()\n#final_df.head()\n"

In [378]:
data['InvoiceDate'].min(), data['InvoiceDate'].max()

(Timestamp('2010-12-01 08:26:00'), Timestamp('2011-12-09 12:50:00'))

### As we can see here we need to segeregate Customers we can directly think of finding Recency, Frequency & Monetory and Based on we can come to conclusion 

* so using GroupBy close and finding RFM ..

#### Lets use RFM recency, frequency, monetary for 
#### customer segmentation using group by Columns like InvoiceDate, InvoiceNo, & Revenue/Bills

In [379]:
import datetime as dt
PRESENT = dt.datetime(2021,7,4)
#print (PRESENT)
#data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])
#data.head()

rfm= data.groupby('CustomerID').agg({'InvoiceDate': lambda date: (PRESENT - date.max()).days,
                                        'InvoiceNo': lambda num: len(num),
                                        'Bill': lambda price: price.sum()})

In [380]:
rfm.columns



Index(['InvoiceDate', 'InvoiceNo', 'Bill'], dtype='object')

In [381]:
# Change the name of columns
rfm.columns=['recency','frequency','monetary']
rfm['recency'] = rfm['recency'].astype(int)
rfm.head()

recency  frequency  monetary
CustomerID                              
12346.0        3819          2      0.00
12347.0        3496        182   4310.00
12348.0        3569         31   1797.24
12349.0        3512         73   1757.55
12350.0        3804         17    334.40

In [382]:
# Devide results in Lowest recency, Highest Frequency and Highest Monetary , based on which we can calculate the score 
rfm['r_quartile'] = pd.qcut(rfm['recency'], 4, ['1','2','3','4'])
rfm['f_quartile'] = pd.qcut(rfm['frequency'], 4, ['4','3','2','1'])
rfm['m_quartile'] = pd.qcut(rfm['monetary'], 4, ['4','3','2','1'])

In [383]:
rfm.head()

recency  frequency  monetary r_quartile f_quartile m_quartile
CustomerID                                                               
12346.0        3819          2      0.00          4          4          4
12347.0        3496        182   4310.00          1          1          1
12348.0        3569         31   1797.24          3          3          1
12349.0        3512         73   1757.55          2          2          1
12350.0        3804         17    334.40          4          4          3

In [384]:
#Aggregate values of Quartile to calculate RFM score 

rfm['RFM_Score'] = rfm.r_quartile.astype(str)+ rfm.f_quartile.astype(str) + rfm.m_quartile.astype(str)
rfm.head()

recency  frequency  monetary r_quartile f_quartile m_quartile  \
CustomerID                                                                  
12346.0        3819          2      0.00          4          4          4   
12347.0        3496        182   4310.00          1          1          1   
12348.0        3569         31   1797.24          3          3          1   
12349.0        3512         73   1757.55          2          2          1   
12350.0        3804         17    334.40          4          4          3   

           RFM_Score  
CustomerID            
12346.0          444  
12347.0          111  
12348.0          331  
12349.0          221  
12350.0          443

In [386]:
# Filter out Top/Best cusotmers

rfm[rfm['RFM_Score']=='111'].sort_values('monetary', ascending=False).head()

recency  frequency   monetary r_quartile f_quartile m_quartile  \
CustomerID                                                                   
14646.0        3495       2085  279489.02          1          1          1   
18102.0        3494        433  256438.49          1          1          1   
17450.0        3502        351  187482.17          1          1          1   
14911.0        3495       5903  132572.62          1          1          1   
14156.0        3503       1420  113384.14          1          1          1   

           RFM_Score  
CustomerID            
14646.0          111  
18102.0          111  
17450.0          111  
14911.0          111  
14156.0          111

### Conclusion: 
* Customer segmentation gets easy using RFM and we can get top customer by getting values of Recency, Frequemcy and Monetary 
* UnitPrice and Units gives us total invoice value for each customers for unique date
* Cancelled order by 'C' initial in Invoice may reduce Monetory
* Description is ignored as it has no impact as of now on customer segmentation as Monery and frequecy did the work for us 
* There are few CustomerID which are not present - it may impact the results but as of now i have neglected